File: kotlin/05_面向对象/01_类和对象.kt
// ============================================================
// Kotlin 面向对象 - 类和对象
// 编译运行：kotlinc 01_类和对象.kt -include-runtime -d out.jar && java -jar out.jar
// ============================================================

fun main() {
    println("=== 1. 类的基础 ===")

    // 基本类
    class Person(val name: String, var age: Int) {
        fun greet() {
            println("你好，我是$name，$age 岁")
        }

        fun haveBirthday() {
            age++
            println("$name 过生日了，现在 $age 岁")
        }
    }

    val p = Person("张三", 25)
    p.greet()
    p.haveBirthday()
    println("修改后年龄: ${p.age}")

    // ============================================================
    // 2. 构造函数
    // ============================================================

    println("\n=== 2. 构造函数 ===")

    // 主构造函数 + init 块
    class User(val name: String, val age: Int) {
        val isAdult: Boolean

        init {
            isAdult = age >= 18
            println("用户 $name 创建完成")
        }

        fun info() {
            println("姓名: $name, 年龄: $age, 成年: $isAdult")
        }
    }

    val user = User("李四", 20)
    user.info()

    // 次构造函数
    class Student {
        val name: String
        val age: Int
        val school: String

        constructor(name: String, age: Int) {
            this.name = name
            this.age = age
            this.school = "未知学校"
        }

        constructor(name: String, age: Int, school: String) {
            this.name = name
            this.age = age
            this.school = school
        }

        fun info() {
            println("$name, $age 岁, $school")
        }
    }

    val s1 = Student("王五", 18)
    val s2 = Student("赵六", 20, "清华大学")
    s1.info()
    s2.info()

    // ============================================================
    // 3. 继承
    // ============================================================

    println("\n=== 3. 继承 ===")

    // open 表示可以被继承
    open class Animal(val name: String) {
        open fun speak() {
            println("$name 发出声音")
        }

        fun eat() {
            println("$name 在吃东西")
        }
    }

    class Dog(name: String) : Animal(name) {
        override fun speak() {
            println("$name 汪汪叫")
        }

        fun fetch() {
            println("$name 在接飞盘")
        }
    }

    class Cat(name: String) : Animal(name) {
        override fun speak() {
            println("$name 喵喵叫")
        }

        fun catchMouse() {
            println("$name 在抓老鼠")
        }
    }

    val dog = Dog("旺财")
    dog.speak()
    dog.fetch()
    dog.eat()

    val cat = Cat("咪咪")
    cat.speak()
    cat.catchMouse()

    // ============================================================
    // 4. 可见性修饰符
    // ============================================================

    println("\n=== 4. 可见性修饰符 ===")

    open class Base {
        public val publicProp = "public"
        protected val protectedProp = "protected"
        private val privateProp = "private"
        internal val internalProp = "internal"

        fun test() {
            println("类内部可访问:")
            println("  public: $publicProp")
            println("  protected: $protectedProp")
            println("  private: $privateProp")
            println("  internal: $internalProp")
        }
    }

    class Derived : Base() {
        fun access() {
            println("子类可访问:")
            println("  public: $publicProp")
            println("  protected: $protectedProp")
            println("  internal: $internalProp")
            // println("  private: $privateProp") // 错误，子类不能访问
        }
    }

    val base = Base()
    base.test()
    val derived = Derived()
    derived.access()
    println("外部可访问:")
    println("  public: ${base.publicProp}")
    println("  internal: ${base.internalProp}")
    // println("  protected: ${base.protectedProp}") // 错误
    // println("  private: ${base.privateProp}") // 错误

    // ============================================================
    // 5. 抽象类
    // ============================================================

    println("\n=== 5. 抽象类 ===")

    abstract class Shape {
        abstract fun area(): Double
        abstract fun perimeter(): Double

        fun info() {
            println("面积: ${area()}, 周长: ${perimeter()}")
        }
    }

    class Circle(val radius: Double) : Shape() {
        override fun area(): Double = Math.PI * radius * radius
        override fun perimeter(): Double = 2 * Math.PI * radius
    }

    class Rectangle(val width: Double, val height: Double) : Shape() {
        override fun area(): Double = width * height
        override fun perimeter(): Double = 2 * (width + height)
    }

    val circle = Circle(5.0)
    println("圆形:")
    circle.info()

    val rectangle = Rectangle(4.0, 3.0)
    println("矩形:")
    rectangle.info()

    // ============================================================
    // 6. 单例对象 object
    // ============================================================

    println("\n=== 6. 单例对象 object ===")

    object Singleton {
        val version = "1.0"
        fun info() {
            println("这是单例对象，版本: $version")
        }
    }

    Singleton.info()
    println("Singleton.version = ${Singleton.version}")

    // ============================================================
    // 7. 伴生对象
    // ============================================================

    println("\n=== 7. 伴生对象 ===")

    class MyClass {
        companion object {
            fun create(): MyClass = MyClass()
            val TAG = "MyClass"

            fun fromValue(value: String): MyClass {
                println("从 $value 创建 MyClass")
                return MyClass()
            }
        }

        fun hello() {
            println("Hello from MyClass")
        }
    }

    val instance = MyClass.create()
    println("伴生对象 TAG: ${MyClass.TAG}")
    val instance2 = MyClass.fromValue("test")
    instance.hello()

    // ============================================================
    // 8. 嵌套类和内部类
    // ============================================================

    println("\n=== 8. 嵌套类和内部类 ===")

    class Outer {
        private val bar = 1

        class Nested {
            fun foo() = 2
        }

        inner class Inner {
            fun foo() = bar
        }
    }

    val nested = Outer.Nested()
    println("嵌套类: ${nested.foo()}")

    val inner = Outer().Inner()
    println("内部类: ${inner.foo()}")

    println("\n=== 运行完成 ===")
}

File: python/05_面向对象/01_类和对象.py
# ============================================================
# Python 面向对象 - 类和对象
# ============================================================

print("=== 1. 类的定义 ===")
class Person:
    """人类"""

    def __init__(self, name, age):
        """构造方法，初始化属性"""
        self.name = name
        self.age = age

    def greet(self):
        """问候方法"""
        print(f"你好，我是{self.name}，今年{self.age}岁")

person1 = Person("张三", 25)
person2 = Person("李四", 30)

person1.greet()
person2.greet()

print("\n=== 2. 属性的访问和修改 ===")
print(f"person1.name = {person1.name}")
print(f"person1.age = {person1.age}")
person1.age = 26
print(f"修改后 person1.age = {person1.age}")

print("\n=== 3. 类变量和实例变量 ===")
class Student:
    school = "北京大学"  # 类变量，所有实例共享

    def __init__(self, name, student_id):
        self.name = name  # 实例变量
        self.student_id = student_id  # 实例变量

s1 = Student("小明", "2023001")
s2 = Student("小红", "2023002")

print(f"s1.school = {s1.school}")
print(f"s2.school = {s2.school}")
print(f"Student.school = {Student.school}")

Student.school = "清华大学"  # 修改类变量
print(f"修改类变量后:")
print(f"s1.school = {s1.school}")
print(f"s2.school = {s2.school}")

print("\n=== 4. 私有属性和方法 ===")
class BankAccount:
    def __init__(self, balance=0):
        self.__balance = balance  # 私有属性（双下划线开头）

    def deposit(self, amount):
        if amount > 0:
            self.__balance += amount
            print(f"存入 {amount} 元")
        else:
            print("存款金额必须大于0")

    def withdraw(self, amount):
        if 0 < amount <= self.__balance:
            self.__balance -= amount
            print(f"取出 {amount} 元")
        else:
            print("取款金额无效")

    def get_balance(self):
        return self.__balance

account = BankAccount(1000)
print(f"初始余额: {account.get_balance()}")
account.deposit(500)
account.withdraw(300)
print(f"当前余额: {account.get_balance()}")

print("\n=== 5. 继承 ===")
class Animal:
    def __init__(self, name):
        self.name = name

    def speak(self):
        print(f"{self.name} 发出声音")

class Dog(Animal):
    def speak(self):
        print(f"{self.name} 汪汪叫")

    def fetch(self):
        print(f"{self.name} 去捡球了")

class Cat(Animal):
    def speak(self):
        print(f"{self.name} 喵喵叫")

dog = Dog("旺财")
cat = Cat("咪咪")

dog.speak()
dog.fetch()
cat.speak()

print("\n=== 6. super() 调用父类方法 ===")
class Employee:
    def __init__(self, name, salary):
        self.name = name
        self.salary = salary

    def work(self):
        print(f"{self.name} 在工作")

class Manager(Employee):
    def __init__(self, name, salary, department):
        super().__init__(name, salary)  # 调用父类构造方法
        self.department = department

    def work(self):
        super().work()  # 调用父类方法
        print(f"{self.name} 管理 {self.department} 部门")

manager = Manager("王经理", 20000, "技术部")
manager.work()

print("\n=== 7. 多重继承 ===")
class Flyable:
    def fly(self):
        print(f"{self.name} 会飞")

class Swimmable:
    def swim(self):
        print(f"{self.name} 会游泳")

class Duck(Animal, Flyable, Swimmable):
    pass

duck = Duck("唐老鸭")
duck.speak()
duck.fly()
duck.swim()

print("\n=== 8. 魔术方法 ===")
class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __str__(self):
        return f"Point({self.x}, {self.y})"

    def __add__(self, other):
        return Point(self.x + other.x, self.y + other.y)

    def __eq__(self, other):
        return self.x == other.x and self.y == other.y

p1 = Point(1, 2)
p2 = Point(3, 4)
p3 = Point(1, 2)

print(f"p1 = {p1}")
print(f"p1 + p2 = {p1 + p2}")
print(f"p1 == p3: {p1 == p3}")
print(f"p1 == p2: {p1 == p2}")
